src/nupic/support/group_by.py

```diff
@@ -0,0 +1,161 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class GroupByGenerator(object):
+  """ A custom generator used by groupby to allow for len() calls on the
+      generator returned.
+  """
+
+  __slots__ = ['iterable', 'beginning', 'length', 'end', 'index']
+
+  def __init__(self, iterable, beginning, length):
+    self.iterable = iterable
+    self.beginning = beginning
+    self.length = length
+    self.end = self.beginning + self.length
+    self.index = 0
+
+
+  def __iter__(self):
+    for i in xrange(self.beginning, self.end):
+      self.index = i
+      yield self.iterable[i]
+
+
+  def __len__(self):
+    return self.length - self.index
+
+
+def groupby(lis, fn):
+  """ A custom implementation of itertools.groupby that doesn't reuse the
+      generator objects, allowing for groupByN to not have to capture the
+      group in a list.
+
+  @param lis (list) the list to perform the groupby on
+  @param fn  (function) the key function to perform the groupings
+
+  @return (tuple) A (key, GroupByGenerator) tuple where the first value is the
+                  key of the group, and second value is the generator that
+                  generates the values in the group.
+  """
+  length = 1
+  beginning = 0
+  l = len(lis)
+
+  if l > 0:
+    key = fn(lis[0])
+    for i in xrange(1, len(lis)):
+      val = fn(lis[i])
```

Comment:
Calling these `key` and `val` can be deceptive since these aren't key-value pairs. Something else would be good, e.g. `key` +`nextKey` or `key` + `key2`.